### DISCUSSION TITLE
Step Analysis / Code Generation

### DISCUSSION THREAD

**Discussion by Code-Grump** [reqnroll/Reqnroll#208] [Ideas]
Because I like to juggle multiple parallel things, and inspired by similar static analysis available for Microsoft's logging system, I thought I would look into adding a Roslyn analyser for Reqnroll steps. Specifically, something that would look at the attributes assigned to a step and verify:

1. The syntax of the text is complete and will compile to a regular expression or Cucumber expression.
2. That the number of matching groups / parameters in the expression matches the number of arguments to be passed to the step method

But then I also got thinking: **why are we compiling step bindings at runtime when we could be doing that at compile-time?** This is how the new [compile-time Regular Expression support](https://learn.microsoft.com/en-us/dotnet/standard/base-types/regular-expression-source-generators?pivots=dotnet-8-0) works, turning a partial method decorated with an attribute into a complete implementation of a regular expression. Likewise, we could turn every step binding attribute into a compiled member that can match and invoke the step.

This means we'd have:
- No need for runtime reflection to invoke steps (because we can generate all that code at runtime)
- No need for parsing step bindings during execution (because we can do all parsing at compile-time)
- Trivial step discovery (because we can generate and expose a collection of them for every assembly)
- An opportunity to pre-compile step-text matching for even faster execution.

There's overlap here; any validation performed by a static analyzer would be somewhat redundant if a code-generator was also going to repeat that work, but both could exist side-by-side.

What would it take to achieve an analyzer?
- Create a new model for parsing step binding text which exposes the metadata needed for parameter analysis. This would have to account for both expression languages we support, ideally expressed as a single model.
- Load that parser into a Roslyn analyzer to indicate step binding syntax errors.

What would it take to achieve code-generation?
- Same new model for parsing step bindings as analyzer
- A code generator to turn step attributes into a registry of generated step classes
- Generalising the runtime to allow runtime-bound and compiled steps to be invoked side-by-side
- An extension to the runtime's discovery of steps to look for compile-time generated step registries exposed by assemblies

Does this seem useful? I certainly like moving runtime concerns to compile-time concerns wherever possible and this seems like a pretty good candidate. The only thing that might make it undesirable is if we eventually wanted to abandon step discovery at runtime completely - **make it so we can only bind to steps that are discoverable at runtime** so that feature code generation would match feature steps to step definitions at compile-time and add direct invocation. There's a lot of drawbacks and challenges to doing so (can't use dynamically loaded assemblies, data-sources have to be available at compile-time, etc), but I know the idea has been brought up as potentially being a long-term goal.

**Comment by gasparnagy** [comment]
Yes, this is certainly a good idea. I was also thinking about this as a long term goal. 

There are (and probably there will be) cases where dynamic binding would be required, but still, I think vast majority of the step-calls are static. When I was thinking about this, my imagined solution was to generate a code that first verifies if dynamic binding is needed (this can be a quick check, for most of the projects as simple as "return false") and if not, than move on with the statically generated stuff.

Nevertheless I am not sure I fully understand the technical solution you summarized. Could you maybe demonstrate this with some simple pseudo-code (ie. what the user needs to do and what do we generate compile time) though a concrete example?

**Reply by Code-Grump** [reply]
Dynamic binding is going to be required for two scenarios that I know of:
1. Whenever an outline is used, our current binding system performs text-replacement on the step template and _then_ attempts to bind it to a matching method. This is fine for inline data, but when data is an external source provided at runtime, we'll be forced to do this dynamically.
2. When steps are defined in assemblies that are **not included in the compilation** - a rare case, but one we technically support, due to the nature of runtime binding. Any number of assemblies can be dropped into the test execution and will be scanned for bindings.

The implementation of the runtime types generated for dynamic invocation would look something like this:

The user creates their step-method as normal:

```
[Binding]
public class AwesomeSteps
{
    [When("I do the thing to {string}")]
    public async Task WhenIDoTheThingTo(string name)
    {
        ...
    }

    [Then("the thing is done")]
    public void ThenTheThingIsDone()
    {
        ...
    }
}
```

The generator observes these decorated methods and adds them to a registry by emitting code like this:

```
internal static class AwesomeStepsDefinitions
{
    // A method to expose all steps
    public static IReadOnlyCollection<IStepDefinition> GetAllStepDefintions() => new List<IStepDefinition>
    { 
        WhenIDoTheThing,
        ThenTheThingIsDone
    }

    // For every step, a class representing the step.
    class WhenIDoTheThingStep : IStepDefinition
    {
        // A dedicated method to invoke the step from an array of arguments.
        public Task InvokeAsync(IObjectContainer container, object[] args)
        {
            return container.Resolve<AwesomeSteps>().WhenIDoTheThingTo((string)args[0]);
        }

        // A dedicated method to check if the step definition is a match for a step in a scenario.
        // The return-type from this would be like a discriminated union, including any matched groups on success. 
        public StepMatchResult MatchStep(string step)
        {
            // If we could, I would invoke a Regex source generator to emit the code needed to execute the expression to
            // avoid any need for runtime compilation.
        }
    }

    public static IScenarioStep WhenIDoTheThing { get; } = new WhenIDoTheThingScenarioStep();

    // Implemented as above, with specifics for the other step.
    class ThenTheThingIsDoneStep : IStepDefinition
    {
        public Task InvokeAsync(IObjectContainer container, object[] args)
        {
            container.Resolve<AwesomeSteps>().ThenTheThingIsDone();
            return Task.CompletedTask;
        }

        ...
    }

    public static IScenarioStep ThenTheThingIsDone { get; } = new ThenTheThingIsDoneScenarioStep();    
```

And finally a class would be included to aggregate everything under a single public operation for ease of discovery.
```
[assembly:StepDefinitionsType(typeof(StepDefinitions))]
public static class StepDefinitions
{
    public static IReadOnlyCollection<IStepDefinition> GetAllStepDefintions() => 
        AwesomeStepsDefinitions.GetAllStepDefinitions
            .Concat(...) // All the other step definitions aggregated under a single discovery point.
            .ToList();
}
```

At runtime, step definitions for a compliant assembly are found in the in that static class - we just need to look for the assembly attribute and expect to find our static type. Absence of the attribute would indicate it doesn't have a registry of step definitions.

The runtime now has a simple model for dynamic step selection, with an opportunity for us to really make that fast. It could be a useful source of information from any tooling that would want to otherwise use reflection.

**Reply by ajeckmans** [reply]
This would fail for anyone dynamically loading steps from assemblies where the step implementations would differ per scenario though. I'm not against doing that (it probably won't work currently either), but we'd better be explicit in the documentation.

Oh nvm. I thought the static class was defined inside of ReqnRoll, but it is the step assembly?

**Reply by Code-Grump** [reply]
This is code that would be added to the compilation of the step assembly.

**Reply by Code-Grump** [reply]
The idea is to generate code that is tailored for each defined step, doing work at compilation-time that would otherwise have to happen at runtime.

**Reply by gasparnagy** [reply]
I really like this concept.

As I mentioned above, this approach does not necessarily incompatible with dynamically loaded/configured step definitions. Those dynamic ones would be stored in a different (runtime) list and the execution could first check the step against the dynamic ones and fall back to the generated ones or something similar. The matching is anyway a regex match so for that it does not matter if the code is generated or not. The problem is more with the invocation by reflection that could be still avoided by the non-dynamic ones.

My idea in addition to this would be the following:

During compilation we could try to match all _known_ step definitions for each step and generate the invocation code to the test method itself. So using your example, let's assume a scenario like this:

```
Scenario: A
When I do the thing to "foo"
Then the thing is done
```

Then the test method we generate could be like:

```
public async Task ScenarioA() {
  // ... setup stuff

  // code for When I do the thing to "foo"
  if (!handle-by-dynamic-step-defs("I do the thing to \"foo\"")) {
    // if there are no dynamic step defs of they do not match:
    var stepMatch = AwesomeStepsDefinitions.WhenDoTheThing.MatchStep("I do the thing to \"foo\"");
    if (stepMatch.Success) {
      await AwesomeStepsDefinitions.WhenDoTheThing.InvokeStep(container, stepMatch.Args);
    }
    else throw...
  }

  // code for Then the thing is done
  if (!handle-by-dynamic-step-defs("Then the thing is done")) {
    // if there are no dynamic step defs of they do not match:
    var stepMatch = AwesomeStepsDefinitions.ThenTheThingIsDone.MatchStep("Then the thing is done");
    if (stepMatch.Success) {
      await AwesomeStepsDefinitions.ThenTheThingIsDone.InvokeStep(container, stepMatch.Args);
    }
    else throw...
  }
}
```

The benefit of this would be that if you then use the "find usages" on the step def method, it would list the feature file lines. (For that we would need to add some line pragmas to the generated step def classes, but we could make it work.) With that the code lens usage list might work as well automatically. In the _worse case_ we could just generate an additional unused ref to the method to convince static code analysis: `await AwesomeStepsDefinitions.WhenDoTheThing.InvokeStep(container, stepMatch.Args, () => AwesomeStepsDefinitions.WhenIDoTheThingTo(null) /* this is not used actually */);`

Of course when we know that there are multiple possible step defs might match (e.g. with scenario outline steps), the code would be a bit more complex, because we would need to invoke all matching (to detect ambiguity) and then invoke the winning one, but something like this would work:

```
  // code for When I do the thing to "foo"
  if (!handle-by-dynamic-step-defs("I do the thing to \"foo\"")) {
    // if there are no dynamic step defs of they do not match:
    var multiStepMatch = HelperClass.MatchStepDefinitions(AwesomeStepsDefinitions.WhenDoTheThing, OtherAwesomeStepsDefinitions.WhenDoTheThing, "I do the thing to \"foo\""); // this would call the match on each and fail if no match or ambiguous 
    switch (multiStepMatch.MatchedIndex) {
      case 0: 
        await AwesomeStepsDefinitions.WhenDoTheThing.InvokeStep(container, multiStepMatch.Args);
        break;
      case 1: 
        await OtherAwesomeStepsDefinitions.WhenDoTheThing.InvokeStep(container, multiStepMatch.Args);
        break;
    }
  }
```

The `HelperClass.MatchStepDefinitions` could also do the dynamic step matching as well of course, so this could be the way to work even if there is only one known matching step def.

**Reply by Code-Grump** [reply]
Just to be clear: this proposes a change to the _runtime_ step-binding such that there's no "parsing" of step definitions; they come "pre-baked" into types and registered in a convenient location so that the runtime just looks for a match and then runs the matching step.

This does pave the way for an easier time doing compile-time binding, such that the invocation of a step is just a trivial method-call. I would expect *no* parsing to be necessary at runtime in these cases, where we have already parsed and extracted parameters at compile-time (save for any custom transformations.)

**Reply by gasparnagy** [reply]
@Code-Grump Yes, I got that. Did I say anything that could be misunderstood and contradicts to this? (For the language clarification, I tend to use the word "parsing" for Gherkin parsing and "binding discovery" or just "discovery" for analyzing the step definition methods and attributes. This latter is what we can avoid doing runtime with this idea.)

**Reply by Code-Grump** [reply]
I would expect the call chain in `TestExecutionEngine.ExecuteStepAsync` to be changed so its search and subsequent step execution are modified to look at the generated step registries to match and then invoke the steps. I would expect this to be controlled by a feature-switch that is all-in: it only looks for generated step definitions or it only performs runtime binding discovery, not a mix of the two.

The generated test-fixtures wouldn't need to change - they already supply everything to perform dynamic invocation. Compile-time binding definitely gets easier on top of this model.
